runtime/compiler/optimizer/InterpreterEmulator.cpp

```diff
@@ -0,0 +1,856 @@
+/*******************************************************************************
+ * Copyright (c) 2000, 2019 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+#include "optimizer/InterpreterEmulator.hpp"
+#include "optimizer/J9EstimateCodeSize.hpp"
+#include "env/VMAccessCriticalSection.hpp"
+#include "optimizer/PreExistence.hpp"
+#include "optimizer/J9CallGraph.hpp"
+#include "ilgen/IlGenRequest.hpp"
+#include "jilconsts.h"
+
+void
+InterpreterEmulator::maintainStackForIf(TR_J9ByteCode bc)
+   {
+   TR_ASSERT_FATAL(_iteratorWithState, "has to be called when the iterator has state!");
+   int32_t branchBC = _bcIndex + next2BytesSigned();
+   int32_t fallThruBC = _bcIndex + 3;
+   IconstOperand * second = pop()->asIconst();
+   IconstOperand * first = pop()->asIconst();
+   bool canBranch = true;
+   bool canFallThru = true;
+   if (second && first)
+      {
+      switch (bc)
+         {
+         case J9BCificmpeq:
+            canBranch = second->intValue == first->intValue;
+            debugTrace(tracer(), "maintainStackForIf ifcmpeq %d == %d\n", second->intValue, first->intValue);
+            break;
+         case J9BCificmpne:
+            canBranch = second->intValue != first->intValue;
+            debugTrace(tracer(), "maintainStackForIf ifcmpne %d != %d\n", second->intValue, first->intValue);
+            break;
```

Comment:
There should be at least some kind of logging or tracing for unhandled bytecodes as well as a static debug counter - if someone changes the bytecodes in the handles that this thing supports we should know about it up front...

Reply:
added assert at method entry to make sure only `J9BCificmpeq` and `J9BCificmpne` are passed in.